### PR TITLE
chore: upgrade libp2p to 0.51

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:
- Upgrade `libp2p` to v0.51
- Refactor and fix logic in `DiscoveryBehaviour`

Manually tested on mainnet, forest caught up to network head successfully and libp2p reelated metrics look good

```
# HELP bitswap_container_capacities Capacity for each bitswap container
# TYPE bitswap_container_capacities gauge
bitswap_container_capacities{type="peer_container_capacity"} 175
bitswap_container_capacities{type="response_channel_container_capacity"} 448
# HELP bitswap_get_block_time Duration of get_block
# TYPE bitswap_get_block_time histogram
bitswap_get_block_time_bucket{le="0.1"} 23912
bitswap_get_block_time_bucket{le="0.5"} 43465
bitswap_get_block_time_bucket{le="0.75"} 65504
bitswap_get_block_time_bucket{le="1"} 66663
bitswap_get_block_time_bucket{le="1.5"} 67484
bitswap_get_block_time_bucket{le="2"} 67752
bitswap_get_block_time_bucket{le="3"} 67926
bitswap_get_block_time_bucket{le="4"} 67956
bitswap_get_block_time_bucket{le="5"} 67956
bitswap_get_block_time_bucket{le="6"} 67956
bitswap_get_block_time_bucket{le="7"} 67956
bitswap_get_block_time_bucket{le="8"} 67956
bitswap_get_block_time_bucket{le="9"} 67956
bitswap_get_block_time_bucket{le="10"} 67956
bitswap_get_block_time_bucket{le="+Inf"} 67956
bitswap_get_block_time_sum 20406.0192123221
bitswap_get_block_time_count 67956
# HELP bitswap_message_count Number of bitswap messages
# TYPE bitswap_message_count counter
bitswap_message_count{type="get_block_success"} 67956
bitswap_message_count{type="inbound_request_have"} 1262
bitswap_message_count{type="inbound_response_block"} 1783807
bitswap_message_count{type="inbound_response_block_not_requested"} 1700656
bitswap_message_count{type="inbound_response_block_update_db"} 67956
bitswap_message_count{type="inbound_response_have_yes"} 908605
bitswap_message_count{type="inbound_stream_count"} 1429695
bitswap_message_count{type="outbound_request_block"} 23902
bitswap_message_count{type="outbound_request_cancel"} 8827798
bitswap_message_count{type="outbound_request_have"} 7243074
bitswap_message_count{type="outbound_stream_count"} 6169013
# HELP bitswap_message_size Size of bitswap messages
# TYPE bitswap_message_size counter
bitswap_message_size{type="inbound_bytes"} 597561808
bitswap_message_size{type="outbound_bytes"} 296064820

# HELP libp2p_messsage_total Total number of libp2p messages by type
# TYPE libp2p_messsage_total counter
libp2p_messsage_total{libp2p_message_kind="chain_exchange_request_in"} 9
libp2p_messsage_total{libp2p_message_kind="chain_exchange_request_out"} 405
libp2p_messsage_total{libp2p_message_kind="chain_exchange_response_in"} 398
libp2p_messsage_total{libp2p_message_kind="chain_exchange_response_out"} 8
libp2p_messsage_total{libp2p_message_kind="hello_request_in"} 354
libp2p_messsage_total{libp2p_message_kind="hello_request_out"} 459
libp2p_messsage_total{libp2p_message_kind="hello_response_in"} 290
libp2p_messsage_total{libp2p_message_kind="hello_response_out"} 354
libp2p_messsage_total{libp2p_message_kind="peer_connected"} 520
libp2p_messsage_total{libp2p_message_kind="peer_disconnected"} 421
libp2p_messsage_total{libp2p_message_kind="pubsub_message_block"} 2269
libp2p_messsage_total{libp2p_message_kind="pubsub_message_message"} 42050

# HELP libp2p_gossipsub_messages Number of messages received.
# TYPE libp2p_gossipsub_messages counter
libp2p_gossipsub_messages_total 44319
# HELP libp2p_identify_protocols Number of connected nodes supporting a specific protocol, with "unrecognized" for each peer supporting one or more unrecognized protocols.
# TYPE libp2p_identify_protocols gauge
libp2p_identify_protocols{protocol="/ipfs/ping/1.0.0"} 390
libp2p_identify_protocols{protocol="/ipfs/id/push/1.0.0"} 390
libp2p_identify_protocols{protocol="unrecognized"} 390
libp2p_identify_protocols{protocol="/ipfs/id/1.0.0"} 390
# HELP libp2p_identify_errors Number of errors while attempting to identify the remote.
# TYPE libp2p_identify_errors counter
libp2p_identify_errors_total 13
# HELP libp2p_identify_pushed Number of times identification information of the local node has been actively pushed to a peer..
# TYPE libp2p_identify_pushed counter
libp2p_identify_pushed_total 0
# HELP libp2p_identify_received Number of times identification information has been received from a peer.
# TYPE libp2p_identify_received counter
libp2p_identify_received_total 5787
# HELP libp2p_identify_received_info_listen_addrs Number of listen addresses for remote peer received in identification information.
# TYPE libp2p_identify_received_info_listen_addrs histogram
libp2p_identify_received_info_listen_addrs_sum 12885.0
libp2p_identify_received_info_listen_addrs_count 5787
libp2p_identify_received_info_listen_addrs_bucket{le="0.0"} 3
libp2p_identify_received_info_listen_addrs_bucket{le="1.0"} 2512
libp2p_identify_received_info_listen_addrs_bucket{le="2.0"} 3519
libp2p_identify_received_info_listen_addrs_bucket{le="4.0"} 5668
libp2p_identify_received_info_listen_addrs_bucket{le="8.0"} 5709
libp2p_identify_received_info_listen_addrs_bucket{le="16.0"} 5762
libp2p_identify_received_info_listen_addrs_bucket{le="32.0"} 5787
libp2p_identify_received_info_listen_addrs_bucket{le="64.0"} 5787
libp2p_identify_received_info_listen_addrs_bucket{le="128.0"} 5787
libp2p_identify_received_info_listen_addrs_bucket{le="256.0"} 5787
libp2p_identify_received_info_listen_addrs_bucket{le="+Inf"} 5787
# HELP libp2p_identify_received_info_protocols Number of protocols supported by the remote peer received in identification information.
# TYPE libp2p_identify_received_info_protocols histogram
libp2p_identify_received_info_protocols_sum 113006.0
libp2p_identify_received_info_protocols_count 5787
libp2p_identify_received_info_protocols_bucket{le="0.0"} 0
libp2p_identify_received_info_protocols_bucket{le="1.0"} 0
libp2p_identify_received_info_protocols_bucket{le="2.0"} 0
libp2p_identify_received_info_protocols_bucket{le="4.0"} 44
libp2p_identify_received_info_protocols_bucket{le="8.0"} 213
libp2p_identify_received_info_protocols_bucket{le="16.0"} 249
libp2p_identify_received_info_protocols_bucket{le="32.0"} 5787
libp2p_identify_received_info_protocols_bucket{le="64.0"} 5787
libp2p_identify_received_info_protocols_bucket{le="128.0"} 5787
libp2p_identify_received_info_protocols_bucket{le="256.0"} 5787
libp2p_identify_received_info_protocols_bucket{le="+Inf"} 5787
# HELP libp2p_identify_sent Number of times identification information of the local node has been sent to a peer in response to an identification request.
# TYPE libp2p_identify_sent counter
libp2p_identify_sent_total 573
# HELP libp2p_identify_listen_addresses Number of listen addresses for remote peer per protocol stack.
# TYPE libp2p_identify_listen_addresses counter
libp2p_identify_listen_addresses_total{protocols="/ip6/tcp"} 103
libp2p_identify_listen_addresses_total{protocols="/ip4/udp/quic-v1"} 84
libp2p_identify_listen_addresses_total{protocols="/dns4/tcp"} 60
libp2p_identify_listen_addresses_total{protocols="/ip4/udp/quic"} 168
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 12466
libp2p_identify_listen_addresses_total{protocols="/dns6/tcp"} 4
# HELP libp2p_kad_query_result_get_record_ok Number of records returned by a successful Kademlia get record query..
# TYPE libp2p_kad_query_result_get_record_ok counter
libp2p_kad_query_result_get_record_ok_total 0
# HELP libp2p_kad_query_result_get_record_error Number of failed Kademlia get record queries..
# TYPE libp2p_kad_query_result_get_record_error counter
# HELP libp2p_kad_query_result_get_closest_peers_ok Number of closest peers returned by a successful Kademlia get closest peers query..
# TYPE libp2p_kad_query_result_get_closest_peers_ok histogram
libp2p_kad_query_result_get_closest_peers_ok_sum 0.0
libp2p_kad_query_result_get_closest_peers_ok_count 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="1.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="2.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="4.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="8.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="16.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="32.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="64.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="128.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="256.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="512.0"} 0
libp2p_kad_query_result_get_closest_peers_ok_bucket{le="+Inf"} 0
# HELP libp2p_kad_query_result_get_closest_peers_error Number of failed Kademlia get closest peers queries..
# TYPE libp2p_kad_query_result_get_closest_peers_error counter
# HELP libp2p_kad_query_result_get_providers_ok Number of providers returned by a successful Kademlia get providers query..
# TYPE libp2p_kad_query_result_get_providers_ok histogram
libp2p_kad_query_result_get_providers_ok_sum 0.0
libp2p_kad_query_result_get_providers_ok_count 0
libp2p_kad_query_result_get_providers_ok_bucket{le="1.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="2.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="4.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="8.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="16.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="32.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="64.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="128.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="256.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="512.0"} 0
libp2p_kad_query_result_get_providers_ok_bucket{le="+Inf"} 0
# HELP libp2p_kad_query_result_get_providers_error Number of failed Kademlia get providers queries..
# TYPE libp2p_kad_query_result_get_providers_error counter
# HELP libp2p_kad_query_result_num_requests Number of requests started for a Kademlia query..
# TYPE libp2p_kad_query_result_num_requests histogram
# HELP libp2p_kad_query_result_num_success Number of successful requests of a Kademlia query..
# TYPE libp2p_kad_query_result_num_success histogram
# HELP libp2p_kad_query_result_num_failure Number of failed requests of a Kademlia query..
# TYPE libp2p_kad_query_result_num_failure histogram
# HELP libp2p_kad_query_result_duration_seconds Duration of a Kademlia query..
# TYPE libp2p_kad_query_result_duration_seconds histogram
# UNIT libp2p_kad_query_result_duration_seconds seconds
# HELP libp2p_kad_routing_updated Number of peers added, updated or evicted to, in or from a specific kbucket in the routing table.
# TYPE libp2p_kad_routing_updated counter
# HELP libp2p_kad_inbound_requests Number of inbound requests.
# TYPE libp2p_kad_inbound_requests counter
# HELP libp2p_ping_rtt_seconds Round-trip time sending a 'ping' and receiving a 'pong'.
# TYPE libp2p_ping_rtt_seconds histogram
# UNIT libp2p_ping_rtt_seconds seconds
libp2p_ping_rtt_seconds_sum 27997.729206133397
libp2p_ping_rtt_seconds_count 102572
libp2p_ping_rtt_seconds_bucket{le="0.001"} 0
libp2p_ping_rtt_seconds_bucket{le="0.002"} 347
libp2p_ping_rtt_seconds_bucket{le="0.004"} 2990
libp2p_ping_rtt_seconds_bucket{le="0.008"} 3399
libp2p_ping_rtt_seconds_bucket{le="0.016"} 3636
libp2p_ping_rtt_seconds_bucket{le="0.032"} 3700
libp2p_ping_rtt_seconds_bucket{le="0.064"} 27084
libp2p_ping_rtt_seconds_bucket{le="0.128"} 39780
libp2p_ping_rtt_seconds_bucket{le="0.256"} 70785
libp2p_ping_rtt_seconds_bucket{le="0.512"} 95691
libp2p_ping_rtt_seconds_bucket{le="1.024"} 98970
libp2p_ping_rtt_seconds_bucket{le="2.048"} 101357
libp2p_ping_rtt_seconds_bucket{le="+Inf"} 102572
# HELP libp2p_ping_failure Failure while sending a 'ping' or receiving a 'pong'.
# TYPE libp2p_ping_failure counter
# HELP libp2p_ping_pong_received Number of 'pong's received.
# TYPE libp2p_ping_pong_received counter
libp2p_ping_pong_received_total 837
# HELP libp2p_swarm_connections_incoming Number of incoming connections per address stack.
# TYPE libp2p_swarm_connections_incoming counter
# HELP libp2p_swarm_connections_incoming_error Number of incoming connection errors.
# TYPE libp2p_swarm_connections_incoming_error counter
# HELP libp2p_swarm_new_listen_addr Number of new listen addresses.
# TYPE libp2p_swarm_new_listen_addr counter
# HELP libp2p_swarm_expired_listen_addr Number of expired listen addresses.
# TYPE libp2p_swarm_expired_listen_addr counter
# HELP libp2p_swarm_listener_closed Number of listeners closed.
# TYPE libp2p_swarm_listener_closed counter
# HELP libp2p_swarm_listener_error Number of listener errors.
# TYPE libp2p_swarm_listener_error counter
libp2p_swarm_listener_error_total 0
# HELP libp2p_swarm_dial_attempt Number of dial attempts.
# TYPE libp2p_swarm_dial_attempt counter
libp2p_swarm_dial_attempt_total 0
# HELP libp2p_swarm_outgoing_connection_error Number outgoing connection errors.
# TYPE libp2p_swarm_outgoing_connection_error counter
# HELP libp2p_swarm_connected_to_banned_peer Number of connection attempts to banned peer.
# TYPE libp2p_swarm_connected_to_banned_peer counter
# HELP libp2p_swarm_connections_established Number of connections established.
# TYPE libp2p_swarm_connections_established counter
# HELP libp2p_swarm_connections_closed Number of connections closed.
# TYPE libp2p_swarm_connections_closed counter
# HELP libp2p_swarm_connections_establishment_duration Time it took (locally) to establish connections.
# TYPE libp2p_swarm_connections_establishment_duration histogram
```

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2597

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: node/forest_libp2p/bitswap/tests/go-app/go.mod [hunk 1]
```diff
@@ -1,22 +1,22 @@
 module test/v2
 
-go 1.18
+go 1.19
```
Comment (hanabi1224) [inline node/forest_libp2p/bitswap/tests/go-app/go.mod:3]: `go-libp2p v0.26.0` has dropped `go 1.18` support

**Review by lemmih — APPROVED** [review]

File: node/forest_libp2p/src/behaviour.rs [hunk 2]
```diff
@@ -95,7 +94,7 @@ impl ForestBehaviour {
             .with_kademlia(config.kademlia)
             .with_user_defined(config.bootstrap_peers.clone())
             // TODO allow configuring this through config.
```
Comment (LesnyRumcajs) [inline node/forest_libp2p/src/behaviour.rs:96]: seems done
```suggestion
```

**Review by LesnyRumcajs — APPROVED** [review]
Faith! More seriously, it'd be fantastic to formalize what "good" metrics mean by enhancing the health check conditions. https://github.com/ChainSafe/forest-iac/blob/main/scripts/sync_check/health_check.sh

File: node/forest_libp2p/src/discovery.rs [hunk 3]
```diff
@@ -215,55 +208,89 @@ impl DiscoveryBehaviour {
...
+    }
 
-        list
+    fn handle_pending_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        maybe_peer: Option<PeerId>,
+        addresses: &[libp2p::Multiaddr],
+        effective_role: libp2p::core::Endpoint,
+    ) -> Result<Vec<libp2p::Multiaddr>, ConnectionDenied> {
```
Comment (aakoshh) [inline node/forest_libp2p/src/discovery.rs:268]: To preserve the previous behaviour of `addresses_of_peer`, shouldn't you concatenate the `user_defined` addresses to the ones returned by Kademlia here?

Comment (hanabi1224) [reply node/forest_libp2p/src/discovery.rs:268]: `user_defined` has been set on [initialization](https://github.com/ChainSafe/forest/pull/2598/files#diff-fa6b31f7c0d7e628fdb5238e78ee8a073b1e52ad1618fb1e08bfa0828651b250R130)

**Pull request merged**